drivers/console/bm_serial.c

```diff
@@ -246,18 +254,23 @@ static void bm_serial_tx_thread(void)
 
 		/* COBS encode frm and then send out */
 		enc_retv = cobs_encode(cobs_encoding_buffer, MAX_BM_FRAME_SIZE, frame_addr, frame_length);
-		//cobs_length = bm_serial_cobs_encode((const uint8_t*) frame_addr, frame_length);
 		//LOG_INF( "Transmitting packet of length: %d", enc_retv.out_len);
 
 		if (enc_retv.status == COBS_ENCODE_OK)
 		{
-			/* Send out UART one byte at a time */
-			i = 0;
-			while (i < enc_retv.out_len)
+			for ( n=0; n < MAX_SERIAL_DEV_COUNT; n++)
 			{
-				uart_poll_out(uart_pipe_dev, cobs_encoding_buffer[i++]);
+				if (upipe_dev[n] != NULL)
+				{
+					/* Send out UART one byte at a time */
```

Comment:
I think this could be changed to write out the entire frame at once.

Reply:
Upon inspection, uart_poll_out() appears to be a blocking API. We probably want to switch to an asynchronous form of serial input/output, either the interrupt driven or DMA approach (which is also required when writing blocks of data, as opposed to byte by byte):
https://docs.zephyrproject.org/latest/hardware/peripherals/uart.html#uart-interrupt-api

Interrupt driven examples:
https://github.com/zephyrproject-rtos/zephyr/blob/main/samples/net/wpan_serial/src/main.c#L367
https://github.com/zephyrproject-rtos/zephyr/blob/main/samples/subsys/usb/hid-cdc/src/main.c